src/code/z_msgevent.c

```diff
@@ -1,105 +1,1229 @@
+    u8* script = *scriptPtr;
+
+    gSaveContext.eventInf[MSCRIPT_GET_8(script, 1)] &= MSCRIPT_GET_8(script, 2) ^ 0xFF;
+    return MSCRIPT_CONTINUE;
+}
+
+/**
+ *
+ * Command structure:
+ *  0:(u8)  cmd
+ *  1:(s16) input player item-action
+ *  3:(s16) skip if player item-action matches input player item-action
+ *  5:(s16) skip if player item-action does not match input player item-action
+ *  7:(s16) skip if player item-action is negative
+ * Command size: 9
+ */
+s32 MsgEvent_BranchOnItemAction(Actor* actor, PlayState* play, u8** scriptPtr, MsgEventCallback callback,
+                                s32* endScript) {
+    u8* script = *scriptPtr;
+    PlayerItemAction checkItemAction = MSCRIPT_GET_16(script, 1);
+    s16 skip;
+    PlayerItemAction curItemAction;
+
+    switch (Message_GetState(&play->msgCtx)) {
+        case TEXT_STATE_CHOICE:
+        case TEXT_STATE_EVENT:
+            if (!Message_ShouldAdvance(play)) {
+                return MSCRIPT_STOP;
+            }
+        case TEXT_STATE_PAUSE_MENU:
+            curItemAction = func_80123810(play);
+
+            if (curItemAction == PLAYER_IA_NONE) {
+                return MSCRIPT_STOP;
+            } else if (curItemAction < 0) {
+                skip = MSCRIPT_GET_16(script, 7);
+            } else if (curItemAction == checkItemAction) {
+                skip = MSCRIPT_GET_16(script, 3);
+            } else {
+                skip = MSCRIPT_GET_16(script, 5);
+            }
+            break;
+
+        default:
+            return MSCRIPT_STOP;
+    }
+    *scriptPtr += skip;
+    return MSCRIPT_CONTINUE;
+}
+
+/**
+ * Branches forward if song is obtained
+ *
+ * Command structure:
+ *  0:(u8)  cmd
+ *  1:(u16) song
+ *  3:(s16) skip
+ * Command size: 9
+ */
+s32 MsgEvent_BranchOnSongObtained(Actor* actor, PlayState* play, u8** scriptPtr, MsgEventCallback callback,
+                                  s32* endScript) {
+    u8* script = *scriptPtr;
+    u16 song = MSCRIPT_GET_16(script, 1);
+    s16 skip = MSCRIPT_GET_16(script, 3);
+
+    if (CHECK_QUEST_ITEM(QUEST_SONG_SONATA + song)) {
+        *scriptPtr += skip;
+    }
+    return MSCRIPT_CONTINUE;
+}
+
+/**
+ * Branches forward if the player's current mask matches
+ *
+ * Command structure:
+ *  0:(u8)  cmd
+ *  1:(s16) mask
+ *  3:(s16) skip
+ * Command size: 9
+ */
+s32 MsgEvent_BranchOnWornMask(Actor* actor, PlayState* play, u8** scriptPtr, MsgEventCallback callback,
+                              s32* endScript) {
+    u8* script = *scriptPtr;
+    s32 mask = MSCRIPT_GET_16(script, 1);
+    s16 skip = MSCRIPT_GET_16(script, 3);
+
+    if (Player_GetMask(play) == mask) {
+        *scriptPtr += skip;
+    }
+    return MSCRIPT_CONTINUE;
+}
+
+#define MSCRIPT_GET_TIME(hour, minute, dest, temp) \
+    (temp) = (hour)*60.0f;                         \
+    (temp) += (minute);                            \
+    (dest) = (temp) * (0x10000 / 60 / 24.0f);      \
+    (dest) = ((dest)-0x10000 / 360 * 90)
```

Comment:
This is the same macro as `SCHEDULE_CALC_TIME`. Maybe we could merge them?

Reply:
Merged the time macros for schedule and msgevent, placed them in macros.h (at least for now), and prefixed them with `SCRIPT` since they seem to deal with scripting languages.

Reply:
Thanks. A bit annoying to have to put them in macros.h, but it should work for now